File: Leetcode/0022 Generate Parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int N) {
        List<string>[] allParenthesis = new List<string>[N + 1];
        allParenthesis[0] = new List<string>(){""};
        for(int n = 1;n <= N;n++)
        {
            allParenthesis[n] = new List<string>();
            for(int k = 0;k < n;k++)
            {
                List<string> inside = allParenthesis[k];
                List<string> right  = allParenthesis[n - 1 - k];
                
                foreach(string i in inside)
                {
                    foreach(string r in right)
                    {
                        allParenthesis[n].Add($"({i}){r}");
                    }
                }
            }
        }
        return allParenthesis[N];
    }
}

File: Leetcode/0022 Generate Parentheses.py
class Solution:
    def generateParenthesis(self, N: int) -> List[str]:
        all_parenthesis = [[] for _ in range(N+1)]
        all_parenthesis[0] = ['']
        for n in range(1,N+1):
            for k in range(n):
                inside = all_parenthesis[k]
                right  = all_parenthesis[n-1-k]
                for i in inside:
                    for r in right:
                        all_parenthesis[n].append(f"({i}){r}")
        return all_parenthesis[N]
        